### PR TITLE
Store newly-uploaded files on save rather than assignment

### PR DESCRIPTION
**Pull request by georgeclaghorn**
At present, when an uploaded file is assigned to a record via a writer method added by `has_one_attached` or `has_many_attached`, it’s persisted to storage immediately. For example, the following causes an uploaded file in `params[:avatar]` to be stored:

```ruby
@user.avatar = params[:avatar]
```

We’d like to support validating blobs attached to records, but for validations to be effective, they ought to prevent invalid files from being stored. It’s of little use to identify an invalid file after it’s already been shipped off to storage: you might use a size validation to limit the cost that a single file can add to your AWS bill, but if the file is stored before validations run, you incur its cost regardless.

This PR changes Active Storage to store attachables assigned to a record after the record is saved rather than immediately. To that end, it introduces change tracking for Active Storage attachments. When you use a writer method added by `has_one_attached` or `has_many_attached`, Active Storage tracks a pending change to the record’s attachments. The change is persisted in the application’s database when the record is saved, and the attachables involved are uploaded to storage only after the save transaction commits.

Included are supplementary reorganization and expansion of the attachment API’s tests.

### CODE REVIEW

File: activestorage/lib/active_storage/attached/macros.rb [hunk 1]
```diff
@@ -34,19 +34,36 @@ def #{name}
...
+          @active_storage_attached_#{name}_change
+        end
+
+        def clear_#{name}_change
+          @active_storage_attached_#{name}_change = nil
         end
       CODE
 
-      has_one :"#{name}_attachment", -> { where(name: name) }, class_name: "ActiveStorage::Attachment", as: :record, inverse_of: :record, dependent: false
+      has_one :"#{name}_attachment", -> { where(name: name) }, class_name: "ActiveStorage::Attachment", as: :record, inverse_of: :record, dependent: :destroy
```
Comment (kaspth) [inline activestorage/lib/active_storage/attached/macros.rb]: Does this need a deprecation cycle?

File: activestorage/lib/active_storage/attached/macros.rb [hunk 2]
```diff
@@ -34,19 +34,36 @@ def #{name}
...
 
-      if dependent == :purge_later
-        after_destroy_commit { public_send(name).purge_later }
-      else
-        before_destroy { public_send(name).detach }
+      after_save { public_send("#{name}_change")&.save }
+
+      after_commit(on: [:create, :update]) do
+        begin
+          public_send("#{name}_change").try(:upload)
```
Comment (kaspth) [inline activestorage/lib/active_storage/attached/macros.rb]: Why not use `&.upload` like above?

File: activestorage/lib/active_storage/attached/one.rb [hunk 3]
```diff
@@ -10,11 +10,11 @@ class Attached::One < Attached
     # You don't have to call this method to access the attachment's methods as
     # they are all available at the model level.
     def attachment
-      record.public_send("#{name}_attachment")
+      change.present? ? change.attachment : record.public_send("#{name}_attachment")
```
Comment (kaspth) [inline activestorage/lib/active_storage/attached/one.rb]: `change&.attachment || record.public_send("#{name}_attachment")`?

File: activestorage/lib/active_storage/attached/one.rb [hunk 4]
```diff
@@ -24,16 +24,10 @@ def blank?
...
 
-      unless blob == blob_was
-        transaction do
-          detach
-          write_attachment build_attachment(blob: blob)
-        end
-
-        blob_was.purge_later if blob_was && dependent == :purge_later
+      if !attached? || new_blob != blob
+        write_attachment build_attachment(blob: new_blob)
```
Comment (kaspth) [inline activestorage/lib/active_storage/attached/one.rb]: Should this write a "change" to the accessor? I don't understand how calling `attach` would now ensure the Attachment got uploaded?

File: activestorage/lib/active_storage/attached/one.rb [hunk 5]
```diff
@@ -69,14 +63,19 @@ def purge
...
     private
       delegate :transaction, to: :record
 
+      def change
+        record.public_send("#{name}_change")
+      end
+
       def build_attachment(blob:)
-        ActiveStorage::Attachment.new(record: record, name: name, blob: blob)
+        Attachment.new(record: record, name: name, blob: blob)
```
Comment (kaspth) [inline activestorage/lib/active_storage/attached/one.rb]: Could also strip `ActiveStorage::` from the change classes.

File: activestorage/app/models/active_storage/attachment.rb [hunk 6]
```diff
@@ -15,17 +15,18 @@ class ActiveStorage::Attachment < ActiveRecord::Base
   delegate_missing_to :blob
 
   after_create_commit :analyze_blob_later, :identify_blob
+  after_destroy_commit :purge_dependent_blob_later
```
Comment (kaspth) [inline activestorage/app/models/active_storage/attachment.rb]: We're switching to `delete` below because of this, right? Is this callback just to ensure that we always honor the `dependent: :destroy` from the macros?

File: activestorage/lib/active_storage/attached/macros.rb [hunk 7]
```diff
@@ -87,11 +104,24 @@ def #{name}
...
+              ActiveStorage::Attached::Changes::CreateMany.new("#{name}", self, attachables)
+            end
+        end
+
+        def #{name}_change
+          @active_storage_attached_#{name}_change
+        end
+
+        def clear_#{name}_change
+          @active_storage_attached_#{name}_change = nil
```
Comment (kaspth) [inline activestorage/lib/active_storage/attached/macros.rb]: It might be nice to replace these ivars with a private/nodoc'ed `attachment_changes` hash in the vein of Active Model's `changes`. Then the change classes could use that and skip the need for `public_send`.

**Review by kaspth — COMMENTED** [review]
Nice! I attempted a cursory review 😄

File: activestorage/lib/active_storage/attached/macros.rb (hunk 2 above)
Comment (georgeclaghorn) [reply activestorage/lib/active_storage/attached/macros.rb]: The `Delete*` changes don’t respond to `upload`. They could implement `upload` and do nothing, and I had them do that originally, but it seemed odd to me.

File: activestorage/lib/active_storage/attached/one.rb (hunk 3 above)
Comment (georgeclaghorn) [reply activestorage/lib/active_storage/attached/one.rb]: Tricky: we don’t want to fall back to the attachment in the DB when a `Delete*` change is pending.

File: activestorage/lib/active_storage/attached/macros.rb (hunk 1 above)
Comment (georgeclaghorn) [reply activestorage/lib/active_storage/attached/macros.rb]: No, it doesn’t change behavior. We did this before by calling `purge_later` or `detach` on destroy.

File: activestorage/lib/active_storage/attached/one.rb (hunk 4 above)
Comment (georgeclaghorn) [reply activestorage/lib/active_storage/attached/one.rb]: We can explore implementing `attach` in terms of the writer in the future. I chose to preserve its existing behavior for now. It uses `create_blob_from`, which uploads as necessary.

File: activestorage/lib/active_storage/attached/macros.rb (hunk 1 above)
Comment (kaspth) [reply activestorage/lib/active_storage/attached/macros.rb]: Got it, so it's just more obvious now.

File: activestorage/lib/active_storage/attached/one.rb (hunk 4 above)
Comment (kaspth) [reply activestorage/lib/active_storage/attached/one.rb]: Ah, got it 👍 There's a fair amount of different flows here 😄

File: activestorage/app/models/active_storage/attachment.rb (hunk 6 above)
Comment (georgeclaghorn) [reply activestorage/app/models/active_storage/attachment.rb]: Yes. This, along with the switch to `dependent: :destroy` on attachment associations, keeps changes from having to track removed attachments and purge them on commit: they can just replace the associated attachments and let AR destroy the removed ones.

File: activestorage/lib/active_storage/attached/macros.rb (hunk 7 above)
Comment (georgeclaghorn) [reply activestorage/lib/active_storage/attached/macros.rb]: Agree. I'll hook that up.

File: activestorage/lib/active_storage/attached/changes/create_one.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      @attachment ||= find_or_build_attachment
+    end
+
+    def blob
+      @blob ||= find_or_build_blob
+    end
+
+    def upload
+      case attachable
+      when ActionDispatch::Http::UploadedFile, Rack::Test::UploadedFile
```
Comment (exviva) [inline activestorage/lib/active_storage/attached/changes/create_one.rb]: Shouldn't we first check if these constants are defined? Are we sure they're guaranteed to be `require`d by now?

Comment (georgeclaghorn) [reply activestorage/lib/active_storage/attached/changes/create_one.rb]: > Are we sure they're guaranteed to be required by now?

Yes. They’re required in `lib/active_storage/attached.rb`.

Comment (exviva) [reply activestorage/lib/active_storage/attached/changes/create_one.rb]: Cool, thanks!

**Comment by yves-vogl** [comment]
Thanks!

**Comment by tobscher** [comment]
@georgeclaghorn That is some excellent work to improve ActiveStorage. It looks like this would potentially fix issue https://github.com/rails/rails/issues/31985.

If this fixes the issue I mentioned before is there any way to benefit from this fix prior to Rails 6?

File: activestorage/lib/active_storage/attached/many.rb [hunk 9]
```diff
@@ -9,7 +9,7 @@ class Attached::Many < Attached
     #
     # All methods called on this proxy object that aren't listed here will automatically be delegated to +attachments+.
     def attachments
-      record.public_send("#{name}_attachments")
+      change.present? ? change.attachments : record.public_send("#{name}_attachments")
```
Comment (glaszig) [inline activestorage/lib/active_storage/attached/many.rb]: this either returns an `Array` or `ActiveStorage::Attachment::ActiveRecord_Associations_CollectionProxy` which makes the interface quite a moving target. with the latter i can do something like `attachments.where(...)` whereas that breaks when getting an array.

**Comment by pjmartorell** [comment]
@georgeclaghorn Is there any plan to add this fix/feature to Rails 5.2 as well? it would be really appreciated :pray:

**Comment by inopinatus** [comment]
This PR broke the usual cycle of validation before save. 

The justification appears to be cost optimisation for folks like Github and Basecamp. 

There is now no official way to access content prior to upload. 

I’m now seeing repeated “wtf?”-level questions as a result in a variety of public and private fora. 

There’s a job to finish here, because out in meat-and-potatoes business apps land where we just want to validate content before save as normal, programmer happiness has been diminished.

**Comment by georgeclaghorn** [comment]
> This PR broke the usual cycle of validation before save.

This PR made it possible to validate file uploads at all. Previously, files were indiscriminately uploaded before validation.

> The justification appears to be cost optimisation for folks like Github and Basecamp.

No.

> There is now no official way to access content prior to upload.

There wasn’t a way to access pre-upload file data in 5.2, either.

> I’m now seeing repeated “wtf?”-level questions as a result in a variety of public and private fora.
>
> There’s a job to finish here, because out in meat-and-potatoes business apps land where we just want to validate content before save as normal, programmer happiness has been diminished.

I’m going to take a look eventually, but I’ve been repeatedly annoyed by your tone on here. Adjust it if you’d like to participate further.

**Comment by inopinatus** [comment]
> There wasn’t a way to access pre-upload file data in 5.2, either.

Just to be clear, since there might be a terminology misunderstanding here:

* By "save", I mean precisely calling `#save` on a model object, and not any more general notion, and definitely not as as a synonym for uploading a blob; and

* By "validate", I mean using Active Model validation callbacks to check the validity of a model's attributes prior to its insertion into the database, and preventing that insertion if validation fails, and including in this the desire to check the validity and acceptability of the content of attachment fields on that model.

In 5.2.3, there was no moment of "pre-upload" at all. As you say, attachments are always uploaded straight away. As a result, `blob#download` always works, and this validation has been working perfectly well in 5.2.3:

```ruby
class ImageValidator < ActiveModel::EachValidator
  def validate_each(record, attribute, value)
    unless value.attached?
      record.errors.add attribute, :blank
      return
    end

    if options[:size]
      record.errors.add attribute, :size unless options[:size] === value.byte_size
    end

    if options[:content]
      record.errors.add attribute, :content unless options[:content] === value.download
    end

    record.errors.add attribute, :unrepresentable unless value.representable?
  end
end

class User < ApplicationRecord
  class ContentCheck
    # Dummy content check for demonstration purposes.
    def ===(content)
      Digest::SHA256.hexdigest(content) =~ /\A[a-f]/
    end
  end

  has_one_attached :avatar
  validates :avatar, image: { size: ...1.megabyte, content: ContentCheck.new }

  #...
end
```

In 6.0.0, this throws an exception in the validator when it hits `value.download`. The context here is super mundane: submitting a form with a file field to to #create action which does `@user = User.new(params)` and then `if @user.save` and so on. As with so many business apps, it really is that dull and ordinary a case.

To put it simply, I am only concerned about having access to the content, whether it is in its final resting place or not, and perceive the loss of that as a regression. It is the model's validation-before-save that matters to us, and not whether or not a loose blob made it to S3 and needs cleaning up later.

I can work around this in a few ways; the simplest, especially for those doing a 5.2 -> 6.0 upgrade, seems to be by using an undocumented internal API, c.f. https://github.com/rails/rails/pull/37005.

**Comment by jbielick** [comment]
It would have been helpful for this information to be in the guides if not the changelog / upgrade guide itself. These changes were breaking and had no deprecation cycle. The work here is beneficial for many I'm sure, but the loop isn't closed on this one.

**Comment by trevorturk** [comment]
@jbielick please do consider improving the documentation if you're able! https://guides.rubyonrails.org/v5.1/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation

**Comment by jbielick** [comment]
@trevorturk The last PR to add some information about this was closed https://github.com/rails/rails/pull/37005

I think closing the loop in the sense that I meant it would be providing a public API to retain existing behavior. Is there a workaround you know could be documented?

**Comment by trevorturk** [comment]
No, I was just reacting to you saying "it would have been helpful for this information to be in the guides" and suggesting you improve the documentation if you're able. If you think changes to the code should be made, you should probably consider opening a new issue referencing this one, or ideally a pull request with your proposed changes.
